8254105: allow static nested declarations

vicente-romero-oracle opened the pull request:
Please review the fix for [JDK-8254105](https://bugs.openjdk.java.net/browse/JDK-8254105). The intention of the fix is to allow static members to be declared inside inner classes. The spec allowing this change can be seen at [Local and Nested Static Declarations](http://cr.openjdk.java.net/~gbierman/8246771/8246771-20200928/specs/local-statics-jls.html). This change is part of the [Records JEP](https://openjdk.java.net/jeps/395). The idea is to allow not only records to be defined inside inner classes but also interfaces, enums, static classes and methods.

TIA,
Vicente

PS: the records spec can be accessed here [Record Classes](http://cr.openjdk.java.net/~gbierman/8246771/8246771-20200928/specs/records-jls.html). This patch also adds a test to check the changes in the Records spec to Chapter 13 "Binary Compatibility". I'm OK moving that test to a separate PR.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8254105](https://bugs.openjdk.java.net/browse/JDK-8254105): allow static nested declarations


### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to 7dfca333e4bbd2034377f77f6991c8ef715f412a


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/571/head:pull/571`
`$ git checkout pull/571`

vicente-romero-oracle commented on test/langtools/tools/javac/LocalInterface.java:
```diff
@@ -2,8 +2,8 @@
  * @bug 8242478
  * @summary test for local interfaces
- * @compile/fail/ref=LocalInterface.out -XDrawDiagnostics LocalInterface.java
- * @compile --enable-preview -source ${jdk.version} LocalInterface.java
```
the current patch is supposed to be applied after the code being reviewed at [PR-290](https://github.com/openjdk/jdk/pull/290) which also modifies this particular line, once [PR-290](https://github.com/openjdk/jdk/pull/290) has been pushed this line and similar should change

vicente-romero-oracle commented on test/langtools/tools/javac/records/RecordsBinaryCompatibilityTests.java:
```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
```
this code is not strictly related to the patch but this test is just basically testing the changes Chapter 13 int the record spec. I'm OK moving this to another PR if needed

vicente-romero-oracle replied:
not really but it is true that it is not strictly related to the patch, will create another issue for it

vicente-romero-oracle commented:
/label remove core-libs

vicente-romero-oracle commented:
Hi David,

This issue is a subtask of JEP 395 along with JDK-8246774. JDK-8246774 is basically removing all the preview annotations from record's code and preview options from the related tests so that records can be used as a standard feature in Java. But issues like this one and probably another I'm still working on related to annotations were, IMO, complex enough to deserve a separate review process. I'm OK with creating an umbrella issue, if that sounds better to you, that contains all implementation tasks and make JDK-8246774 and this current task subtasks of that umbrella issue.

Thanks,
Vicente

vicente-romero-oracle commented:
/csr

mcimadamore reviewed (commented):
The changes look good, but I have one concern: this patch seem to apply the improvements unconditionally, regardless of the source version - is this what we want? While I understand that the risk is low (this patch makes space of compilable programs bigger), at the same times it feels wrong that when compiling with `--source 14` I can now nest static classes inside member classes (which is not compatible with what the SE 14 spec says).

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:
```diff
@@ -1217,24 +1217,19 @@ else if ((sym.owner.flags_field & INTERFACE) != 0)
                     mask |= STATIC;
                 else if ((flags & ENUM) != 0 || (flags & RECORD) != 0) {
-                    log.error(pos, Errors.StaticDeclarationNotAllowedInInnerClasses);
+                    //log.error(pos, Errors.StaticDeclarationNotAllowedInInnerClasses);
```
oops, sure

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:
```diff
@@ -1307,7 +1302,10 @@ else if ((sym.kind == TYP ||
                                 SEALED,
-                           FINAL | NON_SEALED)) {
+                           FINAL | NON_SEALED)
+                 && checkDisjoint(pos, flags,
```
sure I will remove it

lahodaj commented:
As @mcimadamore, I think this should only be allowed for -source 16+, right?

vicente-romero-oracle commented:
> As @mcimadamore, I think this should only be allowed for -source 16+, right?

right but the new goodies are only available if records are allowed. The code is behind `allowRecords` which should only be true for source >= 16, shouldn't that be enough?

vicente-romero-oracle commented:
> As @mcimadamore, I think this should only be allowed for -source 16+, right?

I agree, I'm working on another iteration to fix the issue, thanks for the comments

vicente-romero-oracle commented:
I have pushed a commit addressing the review comments so far. I have re-added code in Attr but this time behind a guard checking if records are allowed in the current source and I have added back some tests I removed before. The tests are now executed with current source and with source 15. Thanks for the comments so far.

mcimadamore reviewed (approved):
Changes look good

vicente-romero-oracle commented:
> Changes look good

thanks!

vicente-romero-oracle commented:
/integrate
